python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -234,3 +255,61 @@ def evaluate_dask(ir: IR) -> DataFrame:
 
     graph, key = task_graph(ir, partition_info)
     return get(graph, key)
+
+
+def _concat(dfs: Sequence[DataFrame]) -> DataFrame:
+    # Concatenate a sequence of DataFrames vertically
+    return Union.do_evaluate(None, *dfs)
+
+
+##
+## Union
+##
+
+
+@lower_ir_node.register(Union)
+def _(
+    ir: Union, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # zlice must be None
+    if ir.zlice is not None:
+        return rec.state["default_mapper"](ir)  # pragma: no cover
+
+    # Lower children
+    children, _partition_info = zip(*(rec(c) for c in ir.children), strict=False)
+    partition_info = reduce(operator.or_, _partition_info)
+
+    # Partition count is the sum of all child partitions
+    count = sum(partition_info[c].count for c in children)
+
+    # Return reconstructed node and partition-info dict
+    new_node = ir.reconstruct(children)
+    partition_info[new_node] = PartitionInfo(count=count)
+    return new_node, partition_info
+
+
+@generate_ir_tasks.register(Union)
+def _(
+    ir: Union, partition_info: MutableMapping[IR, PartitionInfo]
+) -> MutableMapping[Any, Any]:
+    part_out = 0
+    key_name = get_key_name(ir)
+    graph: MutableMapping[Any, Any] = {}
+    for child in ir.children:
+        for i in range(partition_info[child].count):
+            graph[(key_name, part_out)] = (get_key_name(child), i)
+            part_out += 1
+    return graph
+
+
+##
+## Other
+##
+
+
+def _register_other():
+    """Register multi-partition dispatch functions."""
+    import cudf_polars.experimental.io  # noqa: F401
+
+
+_register_other()
```

Comment:
@wence- - I want to avoid stuffing everything in this `parallel.py` file (since there is definitely a lot more logic on the way). Does this seem like a reasonable way to make sure the dispatch functions from those "other" detail files get registered?

Reply:
Okay, but I still need to do all this within `_register_other`, no?

`...io._lower_dataframescan` will depend on definitions in this file.

Reply:
Can we figure out a structure that doesn't have circular imports?

Reply:
Yeah - Agree it probably makes sense to do this refactoring now to keep things "clean" as we add new modules.

We just need to separate out the base logic (e.g. `PartitionInfo`, `get_key_name`, etc...) into a dedicated `base.py` file, and the top-level dispatch functions (e.g. `lower_ir_node` and `generate_ir_tasks`) into a dedicated `dispatch.py` file. This way, both `parallel.py` and `io.py` can import from those files without worrying about circular dependencies.